DataStructures/AATree/AATree.cs

```diff
@@ -0,0 +1,552 @@
+
+            return thisNodeResult && ValidateTree(node.Left) && ValidateTree(node.Right);
+        }
+
+        /// <summary>
+        /// Checks if node is a leaf, and if so if its level is 1.
+        /// </summary>
+        /// <param name="node">The node to check.</param>
+        /// <returns>true if node passes check, false otherwise.</returns>
+        private bool CheckLeafNode(AATreeNode<TKey> node)
+        {
+            bool condition = node.Left == null && node.Right == null && node.Level != 1;
+            return !condition;
+        }
+
+        /// <summary>
+        /// Checks if left node's level is exactly one less than node's level.
+        /// </summary>
+        /// <param name="node">The node to check.</param>
+        /// <returns>true if node passes check, false otherwise.</returns>
+        private bool CheckLeftSubtree(AATreeNode<TKey> node)
+        {
+            bool condition = node.Left != null && node.Level - node.Left.Level != 1;
+            return !condition;
+        }
+
+        /// <summary>
+        /// Checks if right node's level is either equal to or one less than node's level.
+        /// </summary>
+        /// <param name="node">The node to check.</param>
+        /// <returns>true if node passes check, false otherwise.</returns>
+        private bool CheckRightSubtree(AATreeNode<TKey> node)
+        {
+            bool condition = node.Right != null &&
+                node.Level - node.Right.Level != 1 &&
+                node.Level != node.Right.Level;
+            return !condition;
+        }
+
+        /// <summary>
+        /// Checks if right grandchild's (right node's right node) level is less than node.
+        /// </summary>
+        /// <param name="node">The node to check.</param>
+        /// <returns>true if node passes check, false otherwise.</returns>
+        private bool CheckRightGrandChild(AATreeNode<TKey> node)
+        {
+            bool condition = node.Right != null &&
+                node.Right.Right != null &&
+                node.Right.Level < node.Right.Right.Level;
+            return !condition;
+        }
+
+        /// <summary>
+        /// Checks if node is not a leaf, and if so if it has two children.
+        /// </summary>
+        /// <param name="node">The node to check.</param>
+        /// <returns>true if node passes check, false otherwise.</returns>
+        private bool CheckNonLeafChildren(AATreeNode<TKey> node)
+        {
+            bool condition = node.Level > 1 && (node.Left == null || node.Right == null);
+            return !condition;
+        }
+
+        /// <summary>
+        /// Remove right-horizontal links and replaces them with left-horizontal links.
+        /// Acccomplishes this by performing a right rotation.
+        /// </summary>
+        /// <param name="node">The node to rebalence from.</param>
+        /// <returns>The rebalenced node.</returns>
+        private AATreeNode<TKey>? Skew(AATreeNode<TKey>? node)
```

Comment:
```suggestion
        private AATreeNode<TKey> Skew(AATreeNode<TKey> node)
```
I suggest not working with nulls here and checking for them at the call site if it's not too cumbersome.